### ISSUE TITLE
[restrict-plus-operands] false positive on types like `{} & string`

### ISSUE THREAD

**Issue by papb** [typescript-eslint/typescript-eslint#2506]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/restrict-plus-operands": "error"
  }
}
```

```TS
declare const a: {} & string;
declare const b: string;
const x = a + b; // Operands of '+' operation must either be both strings or both numbers. Consider using a template literal. (@typescript-eslint/restrict-plus-operands)
```

**Expected Result**

No error

**Actual Result**

```
Operands of '+' operation must either be both strings or both numbers. Consider using a template literal. (@typescript-eslint/restrict-plus-operands)
```

**Additional Info**

The minimal example above might seem nonsense, using `{} & string`, but it's to be minimal. I encountered a situation in real code involving complex type inferences in which one of the operands apparently ended up being inferred as `string | (Buffer & string)` (which also seems nonsense, maybe I should open an issue at TS too... Regardless, `WhateverType & string` is a string, so the linter shouldn't complain).

**Versions**

| package                            | version (attempt 1, with [XO](https://github.com/xojs/xo)) | version (attempt 2, minimal, without [XO](https://github.com/xojs/xo)) |
| ---------------------------------- | ------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.10.1` | `4.0.1` |
| `@typescript-eslint/parser`        | `3.10.1` | `4.0.1` |
| `TypeScript`                       | `4.0.2` | `4.0.2` |
| `ESLint`                           | `7.7.0` | `7.8.1` |
| `node`                             | `12.14.1` | `12.14.1` |

**Comment by bradzacher** [comment]
> Regardless, WhateverType & string is a string, so the linter shouldn't complain).

That's not strictly true.
`boolean & string === never`
`number & string === never`
`symbol & string === never`
`object & string === never`
`never & string === never`
`any & string === any`

The only cases that it holds true are:
`unknown & string === string`
`string & string === string`
`'string literal' & string === 'string literal'`

The cases of `{ someObjectType } & string` can dubiously be considered strings.

This rule just hasn't had it added.
Happy to accept a PR.

There's some prior art in the `restrict-template-expressions` rule.

https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/restrict-template-expressions.ts

**Comment by sunghyunjo** [comment]
I will try this!

**Comment by papb** [comment]
@sunghyunjo Awesome!!

**Comment by sunghyunjo** [comment]
@bradzacher 
I have one question. I think there is no rule that restricts the intersection type to string, so I think this rule needs to be added. 
ex) create `rules/restrict-intersection-types`

Is this correct approach, or does it need to be fixed by modifying the existing `restrict-plus-operands`?

I am suddenly confused and ask a question. Sorry for the delay.

**Comment by bradzacher** [comment]
A new rule isn't a good idea to solve this problem. The problem isn't that people are able to create these branded types, the problem is that this rule doesn't support them.

Supporting them should be pretty straightforward. You can use restrict-template-expressions for inspiration here.

https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/restrict-template-expressions.ts

It has handling built in for branded types and does similar checks to this rule.
